jenkins/Jenkinsfile

```diff
@@ -0,0 +1,246 @@
+              dir("${env.WORKSPACE}/build_agent") {
+                echo 'install packages'
+                sh '''
+                  curl -O https://bootstrap.pypa.io/get-pip.py
+                  python get-pip.py
+                  yum install -y python-devel git gcc gcc-c++
+                  '''
+                sh script: 'pip install https://github.com/cloudify-cosmo/cloudify-agent-packager/archive/master.zip', label: 'installing agent packager'
+                echo 'creating the agent packager'
+                sh script: '''
+                  cfy-ap -c packaging/local.ini -f -v
+                  '''
+              }
+            }
+          }
+        }
+        stage('Deploy EC2 instances'){
+          steps{
+            catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE'){
+              container('py27'){
+                echo 'Setup Github SSH key'
+                setupGithubSSHKey()
+                dir("${env.WORKSPACE}/${env.PROJECT}/jenkins"){
+                  withVault([configuration: configuration, vaultSecrets: secrets]){
+                    sh script:"""#!/bin/bash
+                      apt-get update
+                      python -m ensurepip --upgrade
+                      python -m pip install --upgrade pip
+                      python -m pip install --upgrade virtualenv
+
+                      virtualenv .venv
+                      source .venv/bin/activate
+
+                      pip install cloudify==5.1.0
+                      cfy profile use ${env.OS_IP} -u ${env.OS_USERNAME} -p ${env.OS_PASSWORD} -t ${env.OS_TENANT}
+                      pushd 'bp'
+                        cfy install -b ec2-py27-blueprint ec2-py27-blueprint.yaml
+                      popd
+                      cfy deployments capabilities ec2-py27-blueprint --json > capabilities.json
+                      echo -e \$(cat capabilities.json | jq '.key_content.value' | tr -d '"') > ~/.ssh/ec2_ssh_key && chmod 600 ~/.ssh/ec2_ssh_key
+                      sleep 120
+                      ssh-keyscan -H \$(cat capabilities.json | jq '.endpoint.value' | tr -d '"') >> ~/.ssh/known_hosts
+                    """, label:'Configure and install blueprint on manager'
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    stage('pytest and build rpm'){
+      parallel {
+        stage('Configure & Run-tests py27 on EC2 instance'){
+          steps{
+            catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE'){
+              container('py27'){
+                dir("${env.WORKSPACE}/${env.PROJECT}/jenkins"){
+                  withVault([configuration: configuration, vaultSecrets: secrets]){
+                    sh script:"""#!/bin/bash
+                      ssh -i ~/.ssh/ec2_ssh_key -l centos \$(cat capabilities.json | jq '.endpoint.value' | tr -d '"') /bin/bash << 'EOT'
+sudo yum install -y epel-release git vim python-virtualenv
```

Comment:
I think you need to create AMI with preinstalled software

Reply:
regarding the AMI, I'm not sure how much that will benefit the process, I prefer those software (especially the git clone and the test/dev requirements) to be up-to-date.
If you think some parts can still be preinstalled in an AMI, let me know which ones and I'll create one